pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
+
+func CreateStrategicMergePatch(a, b []byte, dataStruct interface{}) ([]byte, error) {
+	t, err := getTagStructType(dataStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	aI := map[string]interface{}{}
+	err = json.Unmarshal(a, &aI)
+	if err != nil {
+		return nil, errBadJSONDoc
+	}
+
+	bI := map[string]interface{}{}
+	err = json.Unmarshal(b, &bI)
+	if err != nil {
+		return nil, errBadJSONDoc
+	}
+
+	patch, err := diffMaps(aI, bI, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(patch)
+}
+
+// Returns a (recursive) strategic merge patch that yields b when applied to a.
+func diffMaps(a, b map[string]interface{}, t reflect.Type) (map[string]interface{}, error) {
+	patch := map[string]interface{}{}
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	for key, bv := range b {
+		fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := forkedjson.LookupPatchMetadata(t, key)
+		if err != nil {
+			return nil, err
+		}
+
+		av, ok := a[key]
+		// value was added
+		if !ok {
+			patch[key] = bv
+			continue
+		}
+
+		// If types have changed, replace completely
+		if reflect.TypeOf(av) != reflect.TypeOf(bv) {
+			patch[key] = bv
+			continue
+		}
+
+		// Types are the same, compare values
+		switch at := av.(type) {
+		case map[string]interface{}:
+			bt := bv.(map[string]interface{})
+			pv, err := diffMaps(at, bt, fieldType)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(pv) > 0 {
+				patch[key] = pv
+			}
+		case string, float64, bool:
+			if !matchesValue(av, bv) {
+				patch[key] = bv
+			}
+		case []interface{}:
+			bt := bv.([]interface{})
+			if fieldPatchStrategy == "merge" {
+				pv, err := diffLists(at, bt, fieldType.Elem(), fieldPatchMergeKey)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(pv) > 0 {
+					patch[key] = pv
+				}
+			} else if !matchesArray(at, bt, t) {
+				patch[key] = bv
+			}
+		case nil:
+			switch bv.(type) {
+			case nil:
+				// Both nil, fine.
+			default:
+				patch[key] = bv
+			}
+		default:
+			panic(fmt.Sprintf("Unknown type:%T in key %s", av, key))
+		}
+	}
+
+	// Now add all deleted values as nil
+	for key := range a {
+		_, found := b[key]
+		if !found {
+			patch[key] = nil
+		}
+	}
+
+	return patch, nil
+}
+
+// Returns true if the array matches (must be json types).
+// As is idiomatic for go, an empty array is not the same as a nil array.
+func matchesArray(a, b []interface{}, t reflect.Type) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	if (a == nil && b != nil) || (a != nil && b == nil) {
+		return false
+	}
+
+	for i := range a {
+		if !matchesValue(a[i], b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Returns true if the values matches (must be json types)
+// The types of the values must match, otherwise it will always return false
+// If two map[string]interface{} are given, all elements must match.
+func matchesValue(av, bv interface{}) bool {
+	switch at := av.(type) {
+	case string:
```

Comment:
I don't think you need explicit primitive casting - see http://play.golang.org/p/1xOcMJJ5fO.